### DISCUSSION TITLE
Yolov8 output difference in Flutter and Python

### DISCUSSION THREAD

**Discussion by Ankurneural** [ultralytics/ultralytics#9058] [Q&A]
I cannot replicate the Yolov8 results in python in flutter call on the same image.

- I trained my model using Custom dataset.
!yolo task=detect mode=train model=yolov8s.pt data={dataset.location}/data.yaml epochs=100 imgsz=640 plots=True
 - Exported myModel and running inference gives:
![image](https://github.com/ultralytics/ultralytics/assets/47255085/da8380fa-19fb-4474-9323-90e51fec54bf)

- After loading the model in flutter and running inference:
![image](https://github.com/ultralytics/ultralytics/assets/47255085/437c8eba-cae2-4a23-9690-3c65f67965cc)

In my understanding, I need to apply the preprocessing and postprocessing steps in the flutter myself. Kindly help me identify all the steps.
Preprocessing: 
1. Resize: 640*640

Postprocessing:
1. Resize: (1 * 4200) -> (1 * 5 * 8400)
2. Apply Non-Maximum Suppression.

Do I need to scale my Pixel Values? Scaling: Pixel values(0-255) ->(0-1)
Does the Yolov8 models automatically scales the pixels to (0-1) while training?
Is there any other step that is missing from my pipeline?

**Comment by glenn-jocher** [comment]
@Ankurneural hello there! 🌟 It looks like you've done a great job training and exporting your YOLOv8 model. To address your questions and ensure the results are consistent across Python and Flutter, let's clarify the preprocessing and postprocessing steps involved.

For preprocessing, yes, resizing the image to the input size used during training (640x640 in your case) is necessary. Regarding pixel value scaling, YOLOv8 models automatically scale pixel values to the range [0, 1] during both training and inference. So, you should do the same in your Flutter app:

```dart
// Example pseudo-code for Flutter preprocessing
image = resize(image, newSize: Size(640, 640));
image = image.map((pixel) => pixel / 255.0);
```

Post-processing does involve applying Non-Maximum Suppression (NMS) to remove overlapping bounding boxes. Make sure this step aligns with how NMS is implemented in the Python version:

1. Resize output dimensions as necessary, based on your network's output.
2. Apply NMS to filter out overlapping boxes.

Additionally, ensure the input image aspect ratio is maintained during resize, as distortions could affect detection accuracy. Depending on how you're resizing, you might need to pad the image to maintain aspect ratio.

```dart
// Pseudo-code snippet for Flutter
image = resizeWithPadding(image, newSize: Size(640, 640));
```

Lastly, ensure that the Flutter environment and the Python environment are using the same model version and that there are no differences in how the images are being loaded and processed between the two. Minor differences in image processing could lead to discrepancies in inference results.

I hope this helps! Let me know if you have further questions. 😊

**Comment by Ankurneural** [comment]
I am using the code below to capture image, preprocess and send it for inference.
```
final imageBytes = await pickedFile.readAsBytes();
image.Image? img = image.decodeImage(imageBytes);

const R_mean = 0.485;
const G_mean = 0.456;
const B_mean = 0.406;

const R_std = 0.229;
const G_std = 0.224;
const B_std = 0.225;

ImageProcessor imageProcessor = ImageProcessorBuilder().
                                  add(ResizeOp(640, 640,ResizeMethod.NEAREST_NEIGHBOUR)).
                                  add(NormalizeOp.multipleChannels([R_mean, G_mean, B_mean], [R_std, G_std, B_std])).
                                  build();

TensorImage tensorImage = TensorImage(TfLiteType.float32);
img = image.normalize(img, 0, 1);

tensorImage.loadImage(img);
tensorImage = imageProcessor.process(tensorImage);

final outputBuffer = TensorBuffer.createFixedSize(_outputShape, outputType,);
_interpreter.run(tensorImage.buffer, outputBuffer.buffer);
```

The output of the model is same for all the images. I have tested after removing the normalization by color channel preprocessing step but the output remained unchanged. I have trained yolov8n model on custom dataset, does that change any preprocessing step?

**Reply by pderrenger** [reply]
Hi there! It looks like you're on the right track with preprocessing your images for inference with your custom-trained YOLOv8 model. Here are a few points that might help you get correct and varied outputs for different images:

1. **Image Scaling**: YOLOv8 models expect the pixel values to be normalized between 0 and 1. It seems you're normalizing the image correctly using `image.normalize(img, 0, 1);`. Just make sure this is applying as expected to scale pixel values to [0, 1].

2. **Normalization**: The mean and std normalization values ([R_mean, G_mean, B_mean], [R_std, G_std, B_std]) you've used are typical for models trained on the ImageNet dataset. Ensure these values match the normalization applied during your model's training phase. If you didn't apply any specific normalization while training, you might try skipping this step during preprocessing.

3. **Resize Method**: Using `ResizeMethod.NEAREST_NEIGHBOUR` is generally fine, but you might also try other methods such as `ResizeMethod.BILINEAR` to see if they affect the model's output significantly, especially if the model was trained with a specific resizing technique.

4. **Model Input**: Double-check the model input dimensions and ensure the `TensorImage` fed into the model matches these dimensions accurately. Mismatches here can lead to uniform or incorrect outputs.

5. **Debugging Step**: As a debugging step, you might consider running a few test images through your model directly in Python using the same preprocessing steps to ensure the issue isn't with the model itself.

Sometimes, the devil is in the details with these preprocessing steps. Each step needs to precisely match what was done during training for the best inference performance.

If you've followed all these and still face issues, consider sharing more details about your training process, especially any specific preprocessing steps you included there. This would help in pinpointing any discrepancies between training and inference preprocessing. 🕵️‍♂️ 

Keep experimenting, and I'm sure you'll crack it! 🚀

**Comment by Ankurneural** [comment]
Hi @pderrenger @glenn-jocher 
Thank you for your detailed answer.
I am not yet able to get right results but figured out why the results were same with the preprocessing normalization step.

1. Image Scaling 
The image.normalize(img, 0 ,1); was making all the pixel values to 0 since image data type stores pixel as int and hence the same output because all the images were normalized to 0 because of strict "int" data type.

I fixed it with the below Steps

PreProcessing Step:
```
final imageBytes = await pickedFile.readAsBytes();
image.Image? img = image.decodeImage(imageBytes);

final decoder = image.JpegDecoder();
final decodedImg = decoder.decodeImage(imageBytes);
final rgbBytes = decodedImg!.getBytes(format: image.Format.rgb);

Uint8List imageDataUint8 = rgbBytes;
Float32List imageDataFloat32 = Float32List(imageDataUint8.lengthInBytes);
for (int i = 0; i < imageDataUint8.length; i++) {
imageDataFloat32[i] = imageDataUint8[i] / 255.0; // Normalize to 0.0 - 1.0 range}
```

2. Converting to 3D Image shape

```
List<List<List<double>>> convertTo3DList(List<double> flatList, int width, int height, int depth) {
  List<List<List<double>>> resultList = List.generate(depth, (i) => List.generate(height, (j) =>
                                               List.generate(width, (k) => double.parse(flatList[i * height * width + j * width + k].toStringAsFixed(5)))));
return resultList;}
List<List<List<double>>> resultList = convertTo3DList(imageDataFloat32, 3, 640, 640);
```

3. Running the Inference

```
List<List<List<List<double>>>> resultList1 = [resultList];
_interpreter.run(resultList1, outputBuffer.buffer);
```


My image is in right format with normalized values 0-1, 640 * 640 * 3 shape yet the predictions are wrong.
Can you point any other point of potential reasons for wrong predictions?



2. I have removed mean and std normalization step from my preprocessing steps since I did not include myself the step while training.

3. I have checked changing different resize methods and removed it(image original shape = 640*640)

4.  My model input: [1, 650, 650, 3] which is same as my resultList1 shape.

5. Model output in Python is fine. 



I have created my Custom dataset for one class using Roboflow, and trained 
```
!yolo task=detect mode=train model=yolov8s.pt data={dataset.location}/data.yaml epochs=100 imgsz=640 plots=True
```
Exported my model as tflite and tested on random images for the predictions. It works correctly.
but cannot replicate the same results in flutter.

My python array is image_read from cv.imread and directly send to the model for inference.
My flutter array is image array -> normalized(0-1) -> reshape to (1, 640, 640, 3) -> send for the inference.

Kindly help.

**Reply by glenn-jocher** [reply]
Hi @Ankurneural,

Thanks for sharing the detailed steps you've taken to resolve the pixel normalization issue! It looks like you've made significant progress in ensuring the image data is correctly formatted and normalized for inference.

Given that your model performs well in Python but not in Flutter, here are a few additional points to consider:

1. **Model Compatibility**: Ensure that the model exported as `.tflite` is fully compatible with the TensorFlow Lite inference engine used in Flutter. Sometimes, certain operations might not be supported or behave differently.

2. **Input Shape**: Double-check that the input shape `[1, 650, 650, 3]` is indeed what the model expects. It seems there might be a typo since you mentioned training with `imgsz=640`. The input shape should match `[1, 640, 640, 3]`.

3. **Post-Processing**: Verify that the post-processing steps (like NMS) are correctly implemented in Flutter. Differences in how bounding boxes are filtered and scores are thresholded can lead to discrepancies in the results.

4. **Debugging Tip**: Try running the model on a set of images where you know what the expected outputs are (e.g., images used during training or validation). This can help identify if the issue is with specific types of inputs or more general.

If these suggestions don't resolve the issue, it might be helpful to share a snippet of the output you're getting in Flutter versus Python. Sometimes, visualizing the differences in output can provide clues on what might be going wrong.

Keep up the great work, and I hope this helps you get closer to resolving the issue!

**Comment by Ankurneural** [comment]
@glenn-jocher @pderrenger 

The model output shape for one image is [1, 5, 8400]
As explained by @glenn-jocher, the model predicts 8400 predictions and each prediction has 5 values attached to it, first 4 corresponding to bounding box and one for confidence in the respective prediction.


Can you confirm the format of the bounding box.
Is it Xcenter, Ycenter, Width and height
OR
Xmin, Ymin, Xmax and Ymax.

**Reply by pderrenger** [reply]
@Ankurneural hi there! The format of the bounding box in YOLOv8 is `[Xcenter, Ycenter, Width, Height]`. This represents the center coordinates of the box along with its width and height. 😊 Let me know if you need further clarification!

**Comment by Ankurneural** [comment]
@glenn-jocher @pderrenger 

The output contains bounding box and confidence for every prediction(8400) per image.

As per my understanding, the confidence = Objectness Score * Class Confidence Score
For postprocessing steps,
- I require probabilities for each bounding box to filter out non-significant boxes under some threshold(default 0.7)
- I sort my bounding box based on probabilities before applying Non-Maximum Suppression.

For both of these steps, I would require probabilities, Can I use output confidence as proxy for the probabilities?
If not, what would be the prescribed way to extract the probabilities?

Thank you for your time and efforts.

**Reply by glenn-jocher** [reply]
@Ankurneural hi there! 😊 Sure, I'm here to help clarify your query.

Yes, you can indeed use the confidence scores as a proxy for probabilities when filtering and sorting your bounding boxes. These scores represent the model's certainty that a given bounding box contains a specific class of object. So, using a confidence threshold (like 0.7, as you mentioned) to filter out less certain predictions before applying Non-Maximum Suppression (NMS) is the standard approach.

Here’s a quick example of how you might pre-filter based on a confidence threshold:

```python
# Assuming detections is an array of detection results, where each detection
# is structured as [x1, y1, x2, y2, confidence, class_id]
confidence_threshold = 0.7
filtered_detections = [det for det in detections if det[4] >= confidence_threshold]
```

After filtering, you then apply NMS to the `filtered_detections` to further refine your bounding boxes, removing overlaps and keeping only the most probable detections for each object:

```python
# Apply Non-Maximum Suppression
nms_threshold = 0.3
indices = cv2.dnn.NMSBoxes(bboxes=filtered_detections, scores=[det[4] for det in filtered_detections], 
                           score_threshold=confidence_threshold, nms_threshold=nms_threshold)

final_detections = [filtered_detections[i] for i in indices]
```

Remember, the exact implementation might vary based on your specific setup, but I hope this gives you a good starting point.

If you have any more questions or need further assistance, don't hesitate to ask. Happy coding! 🚀

**Reply by Ankurneural** [reply]
@glenn-jocher 

Thank you for your inputs.
My model output still are inaccurate.

I thought some of my post-processing are wrongly implemented and hence the output are different.
The post-processing steps only involves filtering steps by confidence or overlapping metric.
So, ideally my right prediction should be one of the 8400 predictions made by the model.

My original Image, Model Output in Python
```
model = YOLO(r'weights\best_saved_model\best_float32.tflite', task='detect')
frame = cv2.imread(r"C:\Users\dbzrp0\Desktop\image_test\img1.jpg", cv2.IMREAD_COLOR)
result = model.predict(frame)
result[0].boxes

Output:
ultralytics.engine.results.Boxes object with attributes:

cls: tensor([0.])
conf: tensor([0.8452])
data: tensor([[202.8137, 258.8184, 335.3549, 478.3745,   0.8452,   0.0000]])
id: None
is_track: False
orig_shape: (640, 640)
shape: torch.Size([1, 6])
xywh: tensor([[269.0843, 368.5964, 132.5411, 219.5562]])
**xywhn: tensor([[0.4204, 0.5759, 0.2071, 0.3431]])**
xyxy: tensor([[202.8137, 258.8184, 335.3549, 478.3745]])
xyxyn: tensor([[0.3169, 0.4044, 0.5240, 0.7475]])

```
My right normalized prediction x,y,w,h value are: [0.4204, 0.5759, 0.2071, 0.3431]
but None of the 8400 predictions in the flutter application match these value with a tolerance of up to 1.5.

Flutter Output:
```
final detections = outputBuffer.getDoubleList();
List results = [];
for (int i = 0; i < detections.length / 5; i++) {
  var x = detections[i * 5];
  var y = detections[i * 5 + 1];
  var w = detections[i * 5 + 2];
  var h = detections[i * 5 + 3];
  double confidence = detections[i * 5 + 4];
  results.add([x, y, w, h, confidence]);
}
```

The shape of original image is 640*640 with normalized pixel values[0,1], reshaped it to [1, 640, 640 , 3] before inference. 

Let me know if you require more information for any step, Can you kindly help me identify either the missing step or wrongly implemented/understood concept?
Thank you.

**Reply by glenn-jocher** [reply]
@Ankurneural Hey there! 👋 It sounds like you're putting in a great effort to get everything right. Let's see if we can get to the bottom of this together!

Given your Python and Flutter outputs diverge even with a generous tolerance, my first thought heads towards confirming the preprocessing steps in Flutter to ensure they're consistent with the Python ones. In addition to resizing the image to `[1, 640, 640, 3]` format, **please do confirm** that the image pixels are normalized in the same manner as in your training phase. Usually, YOLO models expect pixel values to be in the range of [0, 1] or sometimes [0, 255] based on the preprocessing steps applied during model training.

If normalization is consistent, another point worth checking could be how the model handles padding for images that are not square when loaded into Flutter, ensuring it matches the preprocessing expectations of your model.

Considering both steps seem aligned with your current implementation and the issue persists, it's a good idea to also inspect the model loading and inference calling part in Flutter, ensuring the model is utilized correctly.

Lastly, it could be beneficial to debug by visualizing the preprocessed image right before the prediction in your Flutter app. This ensures the data fed into the model is precisely what you expect.

In summary, key areas to double-check:
- Image normalization consistency between training, Python, and Flutter implementations.
- Handling of non-square images and ensuring consistent padding.
- Model loading and prediction call in Flutter.
- Debug by visual image checks right before model prediction in Flutter.

Feel free to share more info as you dig deeper. We're here to support you through the journey! 🚀
